compiler/rustc_mir_build/src/thir/pattern/mod.rs

```diff
@@ -562,7 +569,20 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
 
         let args = self.typeck_results.node_args(id);
         let c = ty::Const::new_unevaluated(self.tcx, ty::UnevaluatedConst { def: def_id, args });
-        let pattern = self.const_to_pat(c, ty, id, span);
+        let subpattern = self.const_to_pat(c, ty, id, span);
+        let pattern = if let hir::QPath::Resolved(None, path) = qpath
+            && path.segments.len() == 1
+        {
+            // We only want to mark constants when referenced as bare names that could have been
+            // new bindings if the `const` didn't exist.
```

Comment:
Hm, I assumed we'd be indiscriminately recording all constant expansions and would filter later. I'm not sure which is best, but at least please mention this choice in the `PatKind` variant.

Reply:
Instead of only setting them in this case we could add another flag to the variant to track this and *not* suggest if the flag is set.

Reply:
That would be an improvement imo. I would still prefer to avoid carrying a boolean in THIR for a single diagnostic suggestion tbh, depends how hard it is to do without the boolean.

Reply:
My preference would be: `ExpandedConst { def_id, subpattern }` and do the `span_to_snippet(subpattern.span).chars().all(is_alphanumeric)` dance for diagnostics.

Reply:
Let me know what you think with the last commit.

Reply:
That's what I had in mind, ty. Could you add the `ExpandedConstant` layer inside `const_to_pat` by any chance? To be sure we don't miss one

Reply:
I looked at `const_to_pat`, it wouldn't be super clean to add it there. I'm therefore ok with the current way of doing it.